logtail/logtail.go

```diff
@@ -287,26 +296,31 @@ func (l *logger) uploading(ctx context.Context) {
 	}
 }
 
-func (l *logger) upload(ctx context.Context, body []byte) (uploaded bool, err error) {
+// upload uploads body to the log server.
+// origlen indicates the pre-compression body length.
+// If origlen is -1, the body is not compressed.
+func (l *logger) upload(ctx context.Context, body []byte, origlen int) (uploaded bool, err error) {
 	req, err := http.NewRequest("POST", l.url, bytes.NewReader(body))
 	if err != nil {
 		// I know of no conditions under which this could fail.
 		// Report it very loudly.
 		// TODO record logs to disk
 		panic("logtail: cannot build http request: " + err.Error())
 	}
-	if l.zstdEncoder != nil {
+	if origlen != -1 {
 		req.Header.Add("Content-Encoding", "zstd")
+		req.Header.Add("Orig-Content-Length", strconv.Itoa(origlen))
 	}
 	req.Header["User-Agent"] = nil // not worth writing one; save some bytes
+	req.Header.Add("Content-Length", strconv.Itoa(len(body)))
```

Comment:
http.NewRequest already does this for known reader types: https://play.golang.org/p/XBuRRz-KdUo

And you don't need to add Content-Length to req.Header; request.ContentLength is used.

Reply:
Well I'll be. I went and grepped for "Content-Length" in the source to see whether it happened automatically and didn't find it...because it's called request.ContentLength. I guess I should have just gone and asked the author of the package. :P